sdk/storage/azure-storage-blob/azure/storage/blob/_shared/avro/avro_io.py

```diff
@@ -0,0 +1,616 @@
+            result = True
+        elif (w_type in schema.PRIMITIVE_TYPES and r_type in schema.PRIMITIVE_TYPES
+              and w_type == r_type):
+            result = True
+        elif (w_type == r_type == 'record' and
+              DatumReader.check_props(writer_schema, reader_schema,
+                                      ['fullname'])):
+            result = True
+        elif (w_type == r_type == 'error' and
+              DatumReader.check_props(writer_schema, reader_schema,
+                                      ['fullname'])):
+            result = True
+        elif w_type == r_type == 'request':
+            return True
+        elif (w_type == r_type == 'fixed' and
+              DatumReader.check_props(writer_schema, reader_schema,
+                                      ['fullname', 'size'])):
+            result = True
+        elif (w_type == r_type == 'enum' and
+              DatumReader.check_props(writer_schema, reader_schema,
+                                      ['fullname'])):
+            result = True
+        elif (w_type == r_type == 'map' and
+              DatumReader.check_props(writer_schema.values,
+                                      reader_schema.values, ['type'])):
+            result = True
+        elif (w_type == r_type == 'array' and
+              DatumReader.check_props(writer_schema.items,
+                                      reader_schema.items, ['type'])):
+            result = True
+
+        # Handle schema promotion
+        elif w_type == 'int' and r_type in ['long', 'float', 'double']:
+            result = True
+        elif w_type == 'long' and r_type in ['float', 'double']:
+            result = True
+        elif w_type == 'float' and r_type == 'double':
+            result = True
+        else:
+            result = False
+        return result
+
+    def __init__(self, writer_schema=None, reader_schema=None):
+        """
+        As defined in the Avro specification, we call the schema encoded
+        in the data the "writer's schema", and the schema expected by the
+        reader the "reader's schema".
+        """
+        self._writer_schema = writer_schema
+        self._reader_schema = reader_schema
+
+    # read/write properties
+    def set_writer_schema(self, writer_schema):
+        self._writer_schema = writer_schema
+
+    writer_schema = property(lambda self: self._writer_schema,
+                             set_writer_schema)
+
+    def set_reader_schema(self, reader_schema):
+        self._reader_schema = reader_schema
+
+    reader_schema = property(lambda self: self._reader_schema,
+                             set_reader_schema)
+
+    def read(self, decoder):
+        if self.reader_schema is None:
+            self.reader_schema = self.writer_schema
+        return self.read_data(self.writer_schema, self.reader_schema, decoder)
+
+    def read_data(self, writer_schema, reader_schema, decoder):
+        # schema matching
```

Comment:
why do we need both writer and reader schema?

Reply:
In most general case we can have two schemas - schema in the file that was used to write it. Reader schema can be supplied by someone trying to parse avro file to narrow down things they care about. It's kind of equivalent of "casting" in strongly type language.

However, I think we are not going to use it in foreseeable future, so I'm removing that and just leaving `writer_schema`.

Please also notice how `writer_schema` is defined:
`As defined in the Avro specification, we call the schema encoded in the data the "writer's schema".`